zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/DocumentController.java

```diff
@@ -0,0 +1,99 @@
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestPart;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+
+@CamundaRestController
+@RequestMapping("/v2/documents")
+public class DocumentController {
+
+  private final DocumentServices documentServices;
+
+  @Autowired
+  public DocumentController(final DocumentServices documentServices) {
+    this.documentServices = documentServices;
+  }
+
+  @PostMapping
+  public CompletableFuture<ResponseEntity<Object>> createDocument(
+      @RequestParam(required = false) String documentId,
+      @RequestParam(required = false) String storeId,
+      @RequestPart("file") MultipartFile file,
+      @RequestPart(value = "metadata", required = false) DocumentMetadata metadata) {
+
+    return RequestMapper.toDocumentCreateRequest(documentId, storeId, file, metadata)
+        .fold(RestErrorMapper::mapProblemToCompletedResponse, this::createDocument);
+  }
+
+  private CompletableFuture<ResponseEntity<Object>> createDocument(
+      final DocumentServices.DocumentCreateRequest request) {
+
+    return RequestMapper.executeServiceMethod(
+        () -> documentServices
+            .withAuthentication(RequestMapper.getAuthentication())
+            .createDocument(request),
+        ResponseMapper::toDocumentReference);
+  }
+
+  @GetMapping("/{documentId}")
+  public CompletableFuture<ResponseEntity<Object>> getDocumentContent(
+      @PathVariable String documentId, @RequestParam(required = false) String storeId) {
+
+    return RequestMapper.executeServiceMethod(
+        () -> documentServices
+            .withAuthentication(RequestMapper.getAuthentication())
+            .getDocumentContent(documentId, storeId),
+        documentContentInputStream -> {
+          final StreamingResponseBody body = documentContentInputStream::transferTo;
+          return new ResponseEntity<>(body, HttpStatus.OK);
+        });
+  }
+
+  @DeleteMapping("/{documentId}")
+  public CompletableFuture<ResponseEntity<Object>> deleteDocument(
+      @PathVariable String documentId, @RequestParam(required = false) String storeId) {
+
+    return RequestMapper.executeServiceMethodWithNoContentResult(
+        () -> documentServices
+            .withAuthentication(RequestMapper.getAuthentication())
+            .deleteDocument(documentId, storeId));
+  }
+
+  @PostMapping("/{documentId}/links")
+  public ResponseEntity<Void> createDocumentLink(
+      @PathVariable String documentId, @RequestBody DocumentLinkRequest linkRequest) {
+
+    // TODO: implement
+    throw new UnsupportedOperationException("Not yet implemented");
```

Comment:
🔧 How about we return an [HTTP 501](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) `ProblemDetail` here?